### PR TITLE
Separated control and compute loop, shorten the critical path, and enable more complicated policies

### PR DESCRIPTION
**Pull request by Ying1123**
Moved from #1182. @xiezhq-hermann @hnyls2002 
(I'm sorry that the original PR has been closed by me accidentally.)

## Motivation
The existing design of the scheduler coupled control logic and model computation. While it simplifies the implementation, the overhead of scheduling can sometimes be non-negligible, especially when interacting with radix tree.
This PR decoupled the control and compute loops into different threads, overlapping computation and tasks that are not necessary on the critical paths. This reduces the scheduling overhead and enables more complicated policies in the future to be implemented on the control plane.
Based on the latest CI benchmark result, the PR would introduce about 10% throughput gain by reducing the overhead on critical path.

## Modifications

1. Serialize access to shared resources including radix tree and memory pools with minimal overhead for safety.
2. A control loop that prepares the next prefill batch, handles finished requests and updates radix tree, concurrently with the compute loop.
3. Making policy scheduler compatible with the new design (to be further refined for corner cases).

## Checklist

- [ ] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/srt/managers/io_struct.py [hunk 1]
```diff
@@ -250,9 +250,9 @@ class BatchTokenIDOut:
     meta_info: List[Dict]
     finished_reason: List[BaseFinishReason]
 
-    def __post_init__(self):
-        # deepcopy meta_info to avoid modification in place
-        self.meta_info = copy.deepcopy(self.meta_info)
+    # def __post_init__(self):
+    #     # deepcopy meta_info to avoid modification in place
+    #     self.meta_info = copy.deepcopy(self.meta_info)
```
Comment (merrymercy) [inline python/sglang/srt/managers/io_struct.py]:
```suggestion
```

File: python/sglang/srt/managers/io_struct.py [hunk 2]
```diff
@@ -250,9 +250,6 @@ class BatchTokenIDOut:
     meta_info: List[Dict]
     finished_reason: List[BaseFinishReason]
 
-    def __post_init__(self):
-        # deepcopy meta_info to avoid modification in place
-        self.meta_info = copy.deepcopy(self.meta_info)
```
Comment (merrymercy) [inline python/sglang/srt/managers/io_struct.py:253]:
```suggestion
```

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 3]
```diff
@@ -77,11 +93,15 @@ def __init__(
         self.prefetch_chunk_size = 512
 
         self.can_use_mem_size = self.size
+        # serializes allocation and deallocation operations to the memory pool
```
Comment (merrymercy) [inline python/sglang/srt/mem_cache/memory_pool.py]:
```suggestion
        # serialize allocation and deallocation operations to the memory pool
```

**Comment by merrymercy** [comment]
What is the status of this PR? Do we still plan to merge it?

**Comment by xiezhq-hermann** [comment]
> What is the status of this PR? Do we still plan to merge it?

I think so, @hnyls2002 did a fix on the case of `tp_size > 1`, which should behave the same as before and serve as a temporary fix. For the single GPU setting, it achieves 10%-15% throughput gain. We can always wait to have a more complete PR to merge, but I would suggest doing it incrementally before it diverges too much.
We might still need some style change before merging so please help review the PR. @merrymercy @Ying1123

**Comment by xiezhq-hermann** [comment]
> What is the status of this PR? Do we still plan to merge it?

I think so, @hnyls2002 did a fix on the case of `tp_size > 1`, which should behave the same as before and serve as a temporary fix. For the single GPU setting, it achieves 10%-15% throughput gain. We can always wait to have a more complete PR to merge, but I would suggest to do it incrementally
We might still need some style change before merging but please help review the PR. @merrymercy @Ying1123

**Comment by merrymercy** [comment]
goto #1677
